liquidity/liquidity.go

```diff
@@ -228,14 +243,49 @@ func (m *Manager) getEligibleChannels(ctx context.Context,
 	var (
 		existingOut = make(map[lnwire.ShortChannelID]bool)
 		existingIn  = make(map[route.Vertex]bool)
+		failedOut   = make(map[lnwire.ShortChannelID]time.Time)
 	)
 
+	// Failure cutoff is the most recent failure timestamp we will still
+	// consider a channel eligible. Any channels involved in swaps that have
+	// failed since this point will not be considered.
+	failureCutoff := m.cfg.Clock.Now().Add(m.params.FailureBackOff * -1)
+
 	for _, out := range loopOut {
 		var (
 			state   = out.State().State
 			chanSet = out.Contract.OutgoingChanSet
 		)
 
+		// If a loop out swap failed due to off chain payment after our
+		// failure cutoff, we add all of its channels to a set of
+		// recently failed channels. It is possible that not all of
+		// these channels were used for the swap, but we play it safe
+		// and back off for all of them.
+		//
+		// We only backoff for off chain payment failures because this
+		// error indicates that our swap failed to route. We do not
+		// want to repeatedly try to route through bad channels which
+		// remain unbalanced because they cannot route a swap, so we
+		// backoff. Our other failure types do not represent temporary
+		// failures, so we do not backoff. We don't backoff on the
+		// temporary state itself because we require a restart to clear
+		// this state, which would prevent use of suggestions
+		// completely.
```

Comment:
Wondering whether we shouldn't just backoff entirely if there's a temporary failure? In the case of a "persistent" temp fail (eg bitcoind being down?), we don't really address the issue by backing off for a particular channel, since temp failures aren't really swap specific?

Reply:
Yes, maybe that is the safest thing to do. Because after a restart, those temporarily failed swaps may also again come to life.